src/callgraphgenerator.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+  callgraphgenerator.cpp
+
+  This file is part of Hotspot, the Qt GUI for performance analysis.
+
+  Copyright (C) 2021 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Author: Lieven Hey <lieven.hey@kdab.com>
+
+  Licensees holding valid commercial KDAB Hotspot licenses may use this file in
+  accordance with Hotspot Commercial License Agreement provided with the Software.
+
+  Contact info@kdab.com if any conditions of this licensing are not clear to you.
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "callgraphgenerator.h"
+
+#include <QTextStream>
+
+QHash<Data::Symbol, QString> writeGraph(QTextStream &stream, const Data::Symbol &symbol, const Data::CallerCalleeResults &results, float thresholdPercent, const QString &fontColor)
+{
+    auto settings = Settings::instance();
+    const auto parentId = QUuid::createUuid().toString(QUuid::Id128);
+
+    stream << "digraph callgraph {\n";
+    stream << QStringLiteral("node [shape=box, fontname=\"monospace\", fontcolor=\"%2\", style=filled, color=\"%1\"]\n")
```

Comment:
this should be a QLatin1String, or better yet just directly stream literals:

```
const auto activeColor = settings->callgraphActiveColor().name();
stream << "node [shape=box, fontname=\"monospace\", fontcolor=\"" << fontColor << "\", style=filled, color=\"" << activeColor << "\"]\n";
if (symbol.prettySymbol.isEmpty())
    stream << "??";
else
    stream << symbol.prettySymbol;
```

and so forth, repeats below

this is super nitpicky obviously, but again I'm trying to teach you here :) the reason why this approach is better: QStringLiteral would be a fat utf16 string, but we don't use it as-is, we always mutate it (via the `.arg()`) - at that point the QL1S would be more efficient. But even that would mean we first allocate a QString just to write it to the stream then. By writing to the stream directly, we can reuse its internal buffer directly, which is optimal